### DISCUSSION TITLE
Any suggestions around getting BadHttpRequestException?

### DISCUSSION THREAD

**Discussion by gep13** [CoreWCF/CoreWCF#510] [Q&A]
Has anyone seen this error before when using CoreWCF?

```
Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException: Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate.
```

I am hesitant to change the `MinRequestBodyDataRate` without knowing the full implications of the change.

This seems to happen when there are lots of requests coming into the service.  We didn't see these when running the same service under WCF on full fat .NET Framework, but once we made the switch to CoreWCF, we started seeing these.

Any ideas?

**Comment by gep13** [comment]
@tibor19 @mconnew is this something that either of you have seen before?  Do you have any suggestions on the best path forward?

Thanks! Any help here would be very much appreciated!

**Reply by mconnew** [reply]
HttpListener (which is what WCF uses when self hosted in ,NET Framework) uses a different algorithm for handling slow clients. It has a timer which gets reset every time any bytes are received, and the default is 2 minutes. WCF left this at the default value, and the documentation is available [here](https://docs.microsoft.com/en-us/dotnet/api/system.net.httplistenertimeoutmanager.entitybody?view=netframework-4.8). It would be an understatement to say that is a lot more relaxed than what Kestrel does.  

The [MinRequestBodyDataRate defaults to 240 bytes/s](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.server.kestrel.core.kestrelserverlimits.minrequestbodydatarate?view=aspnetcore-6.0), with a 5 second grace period at the start. 

The risk of setting that value too high is a malicious client can trickle in data at a slow rate consuming resources. The default streaming mode for WCF is buffered, which means WCF accumulates the received data into a buffer until it has the entire request. A malicious client can send up to MaxReceivedMessageSize bytes * number of simultaneous clients allowed amount of RAM. For example, if your max received message size is 1 Megabyte, a client could send a SOAP message which is exactly 1MB big. Sending 240 byte/s a second would take almost 73 minutes to send the message. When a Content-Length header is sent, WCF pre-emptively allocates a buffer that size (as long as not more than MaxReceivedMessageSize). This means a malicious client can force you to allocate 1MB per connection with a low cost on the client side. To stop this sort of trickle attack, you require a minimum amount of data to be received per second.

There are multiple ways to work out what value you want. Which way you choose might be affected by what's causing your problem. Are you having a problem because of a slow client, e.g. clients communicating using cellular such as an Android or iOS app? Or is it a problem of network saturation/contention? One approach might be to take the maximum message size and divide by the expected client timeout period (defaults to 1 minute with WCF) to calculate the rate to set. The problem you can have though is if the connection is saturated and packets are getting dropped, it can take a little while for the bytes to be resent, during which time your average can drop significantly at the start of the request. I suspect your problem is exactly this and that a noisy network is causing some connections to drop early on as they haven't received enough data to not be affected by inconsistent receives at the start of the request.

To be honest, 240 bytes/s is already a super low rate so I doubt you would need to lower it. My suggestion is that you increase the grace period, but keep the rate the same. If you want the same behavior as WCF self-hosted has initially, setting the grace period to 2 minutes would provide that. That does seem too lax to me though. Personally I would try setting the grace period to 20 seconds and see if that improves things, and if it doesn't, then investigate further what's causing it. Maybe enable logging for start of a request and for disconnects and then check your logs to see how long it is before a client gets disconnected because of this. It might give you some insight into whether it's because the grace period is too short, or if the clients have problems later.

Are you sure this wasn't happening on .NET Framework and that you just weren't seeing it? ASP.NET Core can be very noisy about things which WCF/.NET just quietly handled as business as usual.

**Reply by gep13** [reply]
@mconnew thank you very much for such a thorough and detailed response, I really appreciate!  This gives me a lot of things to think about. I will report back if/when we get to a resolution on this. Thanks!
